### PR TITLE
2ds20

### PR DESCRIPTION
**Pull request by manuxi**
Please pull.

### CODE REVIEW

File: src/Message/CompletePurchaseRequest.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    protected function getResourcePath()
+    {
+        return $this->getParameter('resourcePath');
+    }
+
+    public function setResourcePath($resourcePath)
+    {
+        return $this->setParameter('resourcePath', $resourcePath);
+    }
```
Comment (DemigodCode) [inline src/Message/CompletePurchaseRequest.php]: Don't know why we need these methods. Please use getEndpointRoute()

File: src/Message/CompletePurchaseRequest.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    protected function getId()
+    {
+        return $this->getParameter('id');
+    }
+
+    public function setId($id)
+    {
+        return $this->setParameter('id', $id);
+    }
```
Comment (DemigodCode) [inline src/Message/CompletePurchaseRequest.php]: This is the Omnipay TransactionReference Method. See e.g. PurchaseResponse.

File: src/Message/NotificationServerRequest.php [hunk 3]
```diff
@@ -35,6 +36,11 @@ public function isValid() {
         return false;
     }
 
+    public function isRejectedTransaction()
+    {
+        return preg_match('/^(000\.000\.|000\.100\.1|000\.[36])/', $this->data['payload']['result']['code']);
```
Comment (DemigodCode) [inline src/Message/NotificationServerRequest.php]: What means rejected

File: src/Message/PurchaseResponse.php [hunk 4]
```diff
@@ -3,15 +3,53 @@
...
+        if($this->isRedirect()) {
+            return $this->getData()['redirect']['url'];
+        }
+        return '';
+    }
+
+    public function getRedirectData()
+    {
+        if($this->isRedirect()) {
+
```
Comment (DemigodCode) [inline src/Message/PurchaseResponse.php]: Remove empty line

File: src/Message/PurchaseResponse.php [hunk 5]
```diff
@@ -3,15 +3,53 @@
...
+    public function getRedirectMethod()
+    {
+        if($this->isRedirect()) {
+            if(isset($this->getData()['redirect']['method']) && strlen($this->getData()['redirect']['method']) > 0) {
+                return $this->getData()['redirect']['method'];
+            }
+            return 'POST';
+        }
+        return '';
```
Comment (DemigodCode) [inline src/Message/PurchaseResponse.php]: It's better to return null, I guess

File: tests/Message/AbstractRequestTest.php [hunk 6]
```diff
@@ -32,4 +31,21 @@ public function testEndpoint()
         $this->assertSame($this->request, $this->request->setEndpoint('abc123'));
         $this->assertSame('abc123', $this->request->getEndpoint());
     }
+
+    public function testGetData()
+    {
+        $this->request->initialize(['card' => $this->getValidCard()]);
+
+        $data = $this->request->getData();
+
```
Comment (DemigodCode) [inline tests/Message/AbstractRequestTest.php]: Isn't it a good idea to test, if getData returns the testmode?

File: tests/Message/AbstractCardPaymentMessageRequestTest.php [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $this->request = Mockery::mock('\Omnipay\VrPayment\Message\AbstractCardPaymentMessageRequest')->makePartial();
+        $this->request->initialize();
+    }
+
+    public function testGetSimulation()
+    {
+        $this->assertSame($this->request, $this->request->setSimulation(true));
+        $this->assertSame(true, $this->request->getSimulation());
+    }
```
Comment (DemigodCode) [inline tests/Message/AbstractCardPaymentMessageRequestTest.php]: That has to be moved to AbstractRequestTest

File: tests/Message/CompleteAuthorizeResponseTest.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    public function testGetTransactionReference() {
+        $this->assertSame('8ac7a49f6c619208016c61ca7b0c7680', $this->response->getTransactionReference());
+    }
+
+    public function testSendSuccess()
+    {
+        $this->assertTrue($this->response->isSuccessful());
+        $this->assertSame($this->response->getData()['paymentType'], 'PA');
+        $this->assertSame($this->response->getData()['resultDetails']['AcquirerResponse'], 'Success');
+        $this->assertSame('Omnipay\VrPayment\Message\CompleteAuthorizeResponse', get_class($this->response));
```
Comment (DemigodCode) [inline tests/Message/CompleteAuthorizeResponseTest.php]: Most of these are already tested in CompletePurchaseRequestTest, so we don't need them here.

File: tests/Message/AbstractRequestTest.php (hunk 6 above)
Comment (manuxi) [reply tests/Message/AbstractRequestTest.php]: getData() does't return testMode.

File: tests/Message/CompleteAuthorizeResponseTest.php (hunk 8 above)
Comment (manuxi) [reply tests/Message/CompleteAuthorizeResponseTest.php]: Test class deleted

File: src/Message/CompletePurchaseRequest.php (hunk 1 above)
Comment (manuxi) [reply src/Message/CompletePurchaseRequest.php]: CompletePurchaseRequest::setResourcePath() is needed (see Helper::initialize());
CompletePurchaseRequest::getEndpointRoute() returns now directly the param 'resourcePath'

**Pull request merged**